src/Raven.Server/Documents/DocumentsStorage.cs

```diff
@@ -727,37 +727,53 @@ public static long ReadLastEtag(Transaction tx)
         }
 
         public IEnumerable<Document> GetDocumentsStartingWith(DocumentsOperationContext context, string idPrefix, string matches, string exclude, string startAfterId,
-            long start, long take, string collection, DocumentFields fields = DocumentFields.All)
+            long start, long take, string collection, Reference<int> skippedResults, DocumentFields fields = DocumentFields.All)
         {
+            int alreadyReturnedDocumentsCount = 0;
+            int lastLoadedDocumentsCount;
             var isAllDocs = collection == Constants.Documents.Collections.AllDocumentsCollection;
             var requestedDataField = fields.HasFlag(DocumentFields.Data);
             if (isAllDocs == false && requestedDataField == false)
                 fields |= DocumentFields.Data;
-
-            foreach (var doc in GetDocumentsStartingWith(context, idPrefix, matches, exclude, startAfterId, start, take, fields: fields))
+            
+            do
             {
-                if (isAllDocs)
+                lastLoadedDocumentsCount = 0;
+                foreach (var doc in GetDocumentsStartingWith(context, idPrefix, matches, exclude, startAfterId, start, take, fields: fields))
```

Comment:
This overload of `GetDocumentsStartingWith` accepts `Reference<long> skip = null` parameter and uses is under the covers. Please pass there the newly added `skippedResults`.

As I see we don't have the usage where we pass `string matches` parameter but we might have one in the future so this way we'll have this case covered properly.